### PR TITLE
Add typedoc and clean up documentation

### PR DESCRIPTION
**Pull request by rasmus-kirk**
## Purpose

Addresses #174

## Changes

- Added typedoc
- Replaced markdown codeblock examples with inlined runnable examples where applicable
- Updated the CI to lint markdown, build typedoc
- Added workflow that deploys typedoc documentation to GH-pages
  - The new workflow can also be used to release in a safer manner, as the release is made in a docker container. The workflow creates a `build-release` artifact that can be used for this purpose
  - Two custom wrapper actions have been added to make GH-artifacts usable (Will take ~30 times longer without these)
- Added example: `getEmbeddedSchema.ts`
- Updated package READMEs
- Added linkchecker to CI

See [this link](https://rasmus-kirk.github.io/concordium-node-sdk-js/index.html) to check out the compiled documentation

Current references to https://rasmus-kirk.github.io/concordium-node-sdk-js/index.html will be replaced by correct links after merge and before release.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: packages/common/src/GRPCClient.ts
Comment (soerenbf) [inline packages/common/src/GRPCClient.ts:727]: Is this meant to be included twice?

File: packages/nodejs/README.md
Comment (soerenbf) [inline packages/nodejs/README.md:17]: TODO: update

File: packages/nodejs/README.md
Comment (soerenbf) [inline packages/nodejs/README.md:50]: TODO: update links

File: packages/rust-bindings/README.md
Comment (soerenbf) [inline packages/rust-bindings/README.md:8]: TODO: update link

File: packages/web/README.md
Comment (soerenbf) [inline packages/web/README.md:8]: TODO: update link

File: packages/web/README.md
Comment (soerenbf) [inline packages/web/README.md:64]: TODO: update link

File: packages/web/README.md
Comment (soerenbf) [inline packages/web/README.md:69]: TODO: update link

File: packages/web/README.md
Comment (soerenbf) [inline packages/web/README.md:27]: Where did the build instructions go?

File: pages/runnable-examples.md
Comment (soerenbf) [inline pages/runnable-examples.md:1]: Maybe we need to include some information here about the need for building the project first?

**Review by soerenbf — COMMENTED** [review]
Nice work! Will be good with this update to the project. Here are a few things I noticed:

File: packages/common/src/cis2/util.ts
Comment (shjortConcordium) [inline packages/common/src/cis2/util.ts:27]: Did this not completely hide the CIS2 stuff from the docs? thinking

File: packages/web/README.md
Comment (shjortConcordium) [reply packages/web/README.md:27]: The build/release instructions are in the root README.

File: packages/common/src/GRPCClient.ts
Comment (rasmus-kirk) [reply packages/common/src/GRPCClient.ts:727]: No

File: .github/workflows/deployment.yml
Comment (rasmus-kirk) [reply .github/workflows/deployment.yml:1]: But the build-artifact is also used for npm-deployment though

File: packages/common/src/cis0.ts
Comment (rasmus-kirk) [reply packages/common/src/cis0.ts:15]: Yes, I forgot to remove the hidden tag, will do

File: examples/common/statements.ts
Comment (rasmus-kirk) [reply examples/common/statements.ts:25]: Because it didn't work with the other code but I suppose it makes the most sense to keep as a comment

File: .markdown-linkcheck.json
Comment (rasmus-kirk) [reply .markdown-linkcheck.json:1]: It's the config for the markdown link checking tool. It means that it doesn't check any links that includes the pattern regex. So links to `*/modules/Common_GRPC_Client.html` will _not_ be checked

File: .github/workflows/deployment.yml
Comment (shjortConcordium) [reply .github/workflows/deployment.yml:1]: Sure :upside_down_face:

File: .markdown-linkcheck.json
Comment (soerenbf) [reply .markdown-linkcheck.json:1]: I think what I meant here is not so much what the contents of the file does, but more what the module using this file actually does. What does "checking links" actually mean?

File: examples/common/statements.ts
Comment (soerenbf) [reply examples/common/statements.ts:26]: "The statement we wish to prove"? Or what are you proposing here @shjortConcordium? 😄

File: examples/common/aliases.ts
Comment (rasmus-kirk) [reply examples/common/aliases.ts:10]: True, blunder on my end, I just wanted consistent tabulation in the examples, but the correct way to have done that would be to just have the block I guess.

The reasoning is that since we include them in the documentation the non-asynchronous examples should ideally have the same tabulation on the typedoc documentation page.

File: pages/runnable-examples.md
Comment (rasmus-kirk) [reply pages/runnable-examples.md:1]: I'll include it both places

File: README.md
Comment (rasmus-kirk) [reply README.md:36]: It was cumbersome with automatic table generation and I don't think it added much. I don't mind undoing if you disagree.

File: examples/common/aliases.ts
Comment (soerenbf) [reply examples/common/aliases.ts:10]: I see.. I think my opinion is we keep the code as simple and to the point as possible, so I'd prefer if we remove the unnecessary IIFE wrappers where possible.

File: packages/common/README.md
Comment (limemloh) [inline packages/common/README.md:6]: Update link here as well

**Review by shjortConcordium — APPROVED** [review]
LGTM, we can't actually run the new workflow before merging, right? :thinking: :smiling_face_with_tear:

**Comment by rasmus-kirk** [comment]
> LGTM, we can't actually run the new workflow before merging, right? thinking smiling_face_with_tear

Not to my knowledge, no. But it ran on my fork :shrug:

**Review by soerenbf — APPROVED** [review]
I think we should roll this out as well. @rasmus-kirk I'm guessing this should be hosted at `https://concordium.github.io/concordium-node-sdk-js`? Do you know who has access to publish stuff here?

**Comment by abizjak** [comment]
> I think we should roll this out as well. @rasmus-kirk I'm guessing this should be hosted at `https://concordium.github.io/concordium-node-sdk-js`? Do you know who has access to publish stuff here?

We usually do this by using github pages and publish there. I have added you as maintainer so you should be able to create this.

We usually publish documentation from the `gh-pages` branch. This is one example https://github.com/Concordium/concordium-grpc-api/blob/main/.github/workflows/docs.yaml of how it is published.

**Comment by soerenbf** [comment]
> We usually do this by using github pages and publish there. I have added you as maintainer so you should be able to create this.
> 
> We usually publish documentation from the `gh-pages` branch. This is one example https://github.com/Concordium/concordium-grpc-api/blob/main/.github/workflows/docs.yaml of how it is published.
> 
> But you need to enable gh-pages first https://github.com/Concordium/concordium-node-sdk-js/settings/pages

Great, thanks a lot 😄 will look into this when it is merged then!

**Pull request merged**
